### PR TITLE
Add data local service integration

### PR DESCRIPTION
**Pull request by pschorf**
## Changes proposed in this PR
- Adds support for fetching data local costs from a remote server

## Why are we making these changes?
Support a server which provides a notion of the cost to run a job on a given host incorporating data locality. This can then optionally be used by the data local fitness calculator to prefer matching jobs on VMs that support data locality.

### CODE REVIEW

File: integration/tests/cook/test_basic.py [hunk 1]
```diff
@@ -2465,8 +2465,44 @@ def test_pools_in_default_limit_response(self):
                 resp = util.get_limit(self.cook_url, limit, user)
                 self.assertFalse('pools' in resp.json())
 
+    @pytest.mark.serial
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: We can skip this test if the `DATA_LOCAL_SERVICE` environment variable is not set.

File: integration/tests/cook/test_basic.py [hunk 2]
```diff
@@ -2465,8 +2465,44 @@ def test_pools_in_default_limit_response(self):
...
     def test_data_local_support(self):
-        uuid, resp = util.submit_job(self.cook_url, data_local=True)
+        max_cost = util.settings(self.cook_url)['data-local-fitness-calculator']['maximum-cost']
+        slaves = util.get_mesos_state(self.mesos_url)['slaves']
+        costs = []
+        for slave in slaves:
+            costs.append({'node': slave['hostname'], 'cost': max_cost})
+        costs[0]['cost'] = 0
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Can we grade the costs: `0.1 * max-cost`, `0.2 * max-cost` and so on? The remaining parts and assertions of the test don't need to change.

File: scheduler/src/cook/mesos/data_locality.clj [hunk 3]
```diff
@@ -1,32 +1,118 @@
...
-           com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+           com.netflix.fenzo.plugins.BinPackingFitnessCalculators
+           java.util.UUID))
+
+(let [data-local-costs (atom {})
+      last-update-time (atom {})]
+
+  (defn reset-data-local-costs! []
+    (reset! data-local-costs {})
+    (reset! last-update-time {}))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: Can we please name this `job-uuid->last-update-time-atom`?

File: scheduler/src/cook/mesos/data_locality.clj [hunk 4]
```diff
@@ -1,32 +1,118 @@
...
+
+  (defn job-ids-to-update
+    "Return the list of job ids to update data locality costs for in the next iteration.
+     The number of jobs is limited by the configured batch size."
+    [db]
+    (let [{:keys [batch-size]} (config/data-local-fitness-config)
+          pending-jobs (->> db
+                            util/get-pending-job-ents
+                            (filter :job/data-local))
+          last-update-time-val @last-update-time
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: Can we please name this `job-uuid->last-update-time`?

File: scheduler/src/cook/mesos/data_locality.clj [hunk 5]
```diff
@@ -1,32 +1,118 @@
...
+            [plumbing.core :as pc]
+            [clojure.tools.logging :as log]
+            [datomic.api :as d])
   (:import com.netflix.fenzo.VMTaskFitnessCalculator
-           com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+           com.netflix.fenzo.plugins.BinPackingFitnessCalculators
+           java.util.UUID))
+
+(let [data-local-costs (atom {})
+      last-update-time (atom {})]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: Should we expose these data on some admin endpoint for easing debugging data locality support?

File: scheduler/src/cook/mesos/data_locality.clj [hunk 6]
```diff
@@ -1,32 +1,118 @@
...
+            [cook.mesos.util :as util]
+            [plumbing.core :as pc]
+            [clojure.tools.logging :as log]
+            [datomic.api :as d])
   (:import com.netflix.fenzo.VMTaskFitnessCalculator
-           com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+           com.netflix.fenzo.plugins.BinPackingFitnessCalculators
+           java.util.UUID))
+
+(let [data-local-costs (atom {})
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: Should we expose these data on some admin endpoint for easing debugging data locality support?

File: scheduler/src/cook/mesos/data_locality.clj [hunk 7]
```diff
@@ -1,32 +1,118 @@
...
+                                    pending-jobs)
+          missing-data (->> (get have-data?->job false [])
+                            (sort-by :job/submit-time)
+                            (map :job/uuid))
+          sorted-have-data (->> (get have-data?->job true [])
+                                (map :job/uuid)
+                                (sort-by last-update-time-val))
+          pending-job-uuids (->> pending-jobs
+                                 (map :job/uuid)
+                                 (into (hash-set)))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: We can use `(pc/map-from-vals :job/uuid pending-jobs)` here and inline it in the next line.

File: scheduler/src/cook/mesos/data_locality.clj [hunk 8]
```diff
@@ -1,32 +1,118 @@
...
+          no-longer-waiting (set/difference (into (hash-set) (keys last-update-time-val))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: We can use `(-> last-update-time-val keys set)` here.

File: scheduler/src/cook/mesos/data_locality.clj [hunk 9]
```diff
@@ -1,32 +1,118 @@
...
+                            (filter :job/data-local))
+          last-update-time-val @last-update-time
+          have-data?->job (group-by (fn [job] (contains? last-update-time-val (:job/uuid job)))
+                                    pending-jobs)
+          missing-data (->> (get have-data?->job false [])
+                            (sort-by :job/submit-time)
+                            (map :job/uuid))
+          sorted-have-data (->> (get have-data?->job true [])
+                                (map :job/uuid)
+                                (sort-by last-update-time-val))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: `sort-by` is not lazy, can we wrap computing this whole value inside a `(when (< (count missing-data) batch-size)`

File: scheduler/src/cook/mesos/data_locality.clj [hunk 10]
```diff
@@ -1,32 +1,118 @@
...
+       :to-remove no-longer-waiting})))
+
+(defn fetch-data-local-costs
+  "Contacts the server to obtain the data local costs for the given job ids"
+  [job-ids]
+  (let [{:keys [cost-endpoint]} (config/data-local-fitness-config)
+        request {:batch (UUID/randomUUID)
+                 :tasks job-ids}
+        _ (log/debug "Updating data local costs for: " job-ids)
+        {:keys [body]} (http/post cost-endpoint {:body (cheshire/generate-string request)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: We should have a unit tests that acts as documentation for the expected POST body and response template.

File: integration/tests/cook/test_basic.py [hunk 11]
```diff
@@ -2465,8 +2465,44 @@ def test_pools_in_default_limit_response(self):
...
+
+        job = util.load_job(self.cook_url, job_uuid)
         self.assertEqual(True, job['data_local'], job)
+
+        instance = util.wait_for_instance(self.cook_url, job_uuid)
+        self.assertEqual(costs[0]['node'], instance['hostname'])
+
+
+    def test_data_local_constraint(self):
+        num_slaves = len(util.get_mesos_state(self.mesos_url)['slaves'])
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Where is this used?

File: integration/tests/cook/test_basic.py [hunk 12]
```diff
@@ -2465,8 +2465,44 @@ def test_pools_in_default_limit_response(self):
...
+        uuid, resp = util.submit_job(self.cook_url, data_local=True)
+        self.assertEqual(201, resp.status_code, resp.text)
+
+        def query_unscheduled():
+            resp = util.unscheduled_jobs(self.cook_url, uuid)[0][0]
+            placement_reasons = [reason for reason in resp['reasons']
+                                 if reason['reason'] == reasons.COULD_NOT_PLACE_JOB]
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: What in the job causes it to not be scheduled?

File: scheduler/src/cook/mesos/data_locality.clj [hunk 13]
```diff
@@ -1,32 +1,118 @@
...
+      (log/debug "Updating data local costs for " to-fetch)
+      (let [new-costs (fetch-data-local-costs to-fetch)]
+        (log/debug "Got updated costs " new-costs)
+        (update-data-local-costs new-costs to-remove)))))
+
+(defn start-update-cycles!
+  [datomic-conn trigger-chan]
+  (log/info "Starting data local service update chan")
+  (util/chime-at-ch trigger-chan
+                    (fn [] (fetch-and-update-data-local-costs (d/db datomic-conn)))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: There are no `log/info` in the happy path that would tell us that the data locality daemon ran and updated something.

File: scheduler/src/cook/mesos/data_locality.clj (hunk 6 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/data_locality.clj]: Can we please name this `job-uuid->host-name->cost-atom`?

Comment (pschorf) [reply scheduler/src/cook/mesos/data_locality.clj]: Done

File: scheduler/src/cook/mesos/data_locality.clj (hunk 5 above)
Comment (pschorf) [reply scheduler/src/cook/mesos/data_locality.clj]: Done

**Pull request closed without merging**